### PR TITLE
Refactoring `hal.executable.target` -> `variant` and simplifying infra for multi-targeting.

### PR DESCRIPTION
**Pull request by benvanik**
The major change here is renaming `hal.executable.target` to `hal.executable.variant`. This allows for `hal.executable.target` to specify target information and variants to be an instantiation for one or more targets. This initial set of changes is mostly just refactoring/code shuffling to make this renaming possible and adding the skeleton for the future targeting logic.

### CODE REVIEW

File: iree/compiler/Dialect/HAL/Target/LLVM/LLVMAOTTarget.cpp [hunk 1]
```diff
@@ -66,15 +66,9 @@ class LLVMAOTTargetBackend final : public TargetBackend {
       : options_(std::move(options)) {}
 
   // NOTE: we could vary these based on the options, such as by arch/etc.
-  std::string name() const override { return "llvm_aot"; }
-  std::string filter_pattern() const override {
-    llvm::Triple targetTriple(options_.targetTriple);
-    if (targetTriple.isWasm()) {
-      return "wasm*";
```
Comment (ScottTodd) [inline iree/compiler/Dialect/HAL/Target/LLVM/LLVMAOTTarget.cpp:73]: putting "wasm" in the filter/deviceID was load bearing in my work to get wasm running, but I guess we can drop that until the wasm work is fully landed

File: iree/compiler/Dialect/HAL/Target/LLVM/LLVMAOTTarget.cpp [hunk 2]
```diff
@@ -471,17 +462,17 @@ void registerLLVMAOTTargetBackends(
...
 
-  static TargetBackendRegistration dylibRegistration("dylib-llvm-aot", [=]() {
+  auto backendFactory = [=]() {
     INIT_LLVM_TARGET(X86)
     INIT_LLVM_TARGET(ARM)
     INIT_LLVM_TARGET(AArch64)
     INIT_LLVM_TARGET(RISCV)
-    return std::make_unique<LLVMAOTTargetBackend>(queryOptions());
-  });
-  static TargetBackendRegistration wasmRegistration("wasm-llvm-aot", [=]() {
```
Comment (ScottTodd) [inline iree/compiler/Dialect/HAL/Target/LLVM/LLVMAOTTarget.cpp:475]: Same here, I was using this target registration for `-iree-hal-target-backends=wasm-llvm-aot`. Fine to drop it and add back later (esp. when there are no tests verifying that it does what we expect)

File: iree/compiler/Dialect/HAL/IR/HALInterfaces.td [hunk 3]
```diff
@@ -71,4 +71,21 @@ def HAL_SizeAwareOp : OpInterface<"SizeAwareOpInterface"> {
...
+  let description = [{
+    An attribute that specifies a compilation target for device executables.
+    Attributes conforming to this interface are queried throughout the HAL
+    transformation pipeline for their target capabilities and to provide
+    translation, linking, and serialization pipelines.
+  }];
+
+  let methods = [
+    InterfaceMethod<
+      [{TODO}],
```
Comment (ScottTodd) [inline iree/compiler/Dialect/HAL/IR/HALInterfaces.td:85]: Want to elaborate on this TODO? :p

**Review by ScottTodd — APPROVED** [review]

File: iree/compiler/Dialect/HAL/Target/MetalSPIRV/MetalSPIRVTarget.cpp [hunk 4]
```diff
@@ -125,7 +125,7 @@ class MetalSPIRVTargetBackend : public SPIRVTargetBackend {
 
     // 5. Add the binary data to the target executable.
     auto binaryOp = executableBuilder.create<IREE::HAL::ExecutableBinaryOp>(
-        targetOp.getLoc(), targetOp.sym_name(),
+        variantOp.getLoc(), variantOp.sym_name(),
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/compiler/Dialect/HAL/Target/MetalSPIRV/MetalSPIRVTarget.cpp:127]: yeah - 'executable target' is one thing, while 'target executable' is another. english and all :)

File: iree/compiler/Dialect/HAL/Target/TargetBackend.cpp [hunk 5]
```diff
@@ -224,18 +224,18 @@ LogicalResult TargetBackend::linkExecutablesInto(
...
   for (auto sourceExecutableOp : sourceExecutableOps) {
-    auto targetOps = llvm::to_vector<4>(
-        sourceExecutableOp.getOps<IREE::HAL::ExecutableTargetOp>());
-    for (auto targetOp : targetOps) {
+    auto variantOps = llvm::to_vector<4>(
+        sourceExecutableOp.getOps<IREE::HAL::ExecutableVariantOp>());
+    for (auto variantOp : variantOps) {
       // Only process targets matching our pattern.
-      if (!matchPattern(targetOp.target_backend_filter(), filter_pattern())) {
+      if (!matchPattern(variantOp.target_backend_filter(), filter_pattern())) {
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/compiler/Dialect/HAL/Target/TargetBackend.cpp]: here it's talking about the target backend (this ambiguity is why I'm renaming it :)

File: iree/compiler/Dialect/HAL/Target/TargetBackend.h [hunk 6]
```diff
@@ -51,37 +51,41 @@ TargetOptions getTargetOptionsFromFlags();
...
-// to at least one `hal.executable.target` for each unique target backend
+// to at least one `hal.executable.variant` for each unique target backend
 // configuration, and for each of those target backends can emit one or more
-// `hal.executable.target` containing the translated contents. Finally, each
+// `hal.executable.variant` containing the translated contents. Finally, each
 // executable target will be serialized into one or more binary formats. The
-// exact contents of the `hal.executable.target` ops is left to the backends and
-// can contain backend-specific nested IR and attributes.
+// exact contents of the `hal.executable.variant` ops is left to the backends
+// and can contain backend-specific nested IR and attributes.
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/compiler/Dialect/HAL/Target/TargetBackend.h:59]: this whole file is going away - it's all out of date

File: iree/hal/device.h [hunk 7]
```diff
@@ -154,10 +154,16 @@ IREE_API_EXPORT iree_hal_allocator_t* iree_hal_device_allocator(
...
+//   hal.device.id :: some-pattern-*
+//   hal.device.feature :: some-pattern-*
+//   hal.device.architecture : some-pattern-*
+//
 // Returned values must remain the same for the lifetime of the device as
 // callers may cache them to avoid redundant calls.
-IREE_API_EXPORT iree_status_t iree_hal_device_query_i32(
-    iree_hal_device_t* device, iree_string_view_t key, int32_t* out_value);
+IREE_API_EXPORT iree_status_t
+iree_hal_device_query_i32(iree_hal_device_t* device, iree_string_view_t ns,
```
[the remark this answers is not in the document]
Comment (benvanik) [reply iree/hal/device.h:165]: `namespace` is a reserved keyword, so I can't use those characters ;) `ns` is a common abbreviation (xml's `xmlns` etc). Suggestions welcome for alternatives; "group"? "category"?

File: iree/compiler/Dialect/HAL/Target/LLVM/LLVMAOTTarget.cpp (hunk 1 above)
Comment (benvanik) [reply iree/compiler/Dialect/HAL/Target/LLVM/LLVMAOTTarget.cpp:73]: this is all changing, and this particular piece is the simplest kind of work that can be done on bringup so not worth preserving when unused (YAGNI)

File: iree/compiler/Dialect/HAL/IR/HALInterfaces.td (hunk 3 above)
Comment (benvanik) [reply iree/compiler/Dialect/HAL/IR/HALInterfaces.td:85]: in the next PR

File: iree/hal/device.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (ScottTodd) [reply iree/hal/device.h:165]: Hah, good point :P can you at least mention it in comments? The `ns` abbreviation seems fine to me in that case.

[the remark this answers is not in the document]
Comment (benvanik) [reply iree/hal/device.h:165]: Renamed to category.

**Review by ScottTodd — APPROVED** [review]

**Pull request merged**